### PR TITLE
Request governor for Okta Management API calls

### LINKED ISSUE #186: Add request throttling to avoid rate limit violations
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Terraform Okta provider causes rate limit violations. This is problematic for users who have apps that rely on certain Okta APIs. A rate limit violation caused by Terraform Okta provider often effectively causes app downtime. Some organizations that are on the highest pricing plan might have significantly higher rate limits than other organizations on lower pricing plans. Because of that this problem might have gone unnoticed thus far.

Example: `terraform plan` and `terraform apply` cause `/api/v1/apps` API endpoint rate limit violation by making lots of these requests:
1. GET /api/v1/apps/{app_id}/groups?limit=200
2. GET /api/v1/apps/{app_id}/users?limit=200
3. GET /api/v1/apps/{app_id}/groups/{group_id}

We got confirmation from Okta team that all of these endpoints share the same underlying `/api/v1/apps` endpoint rate limit.

Imagine for a moment that your app uses Okta API endpoints that share the `/api/v1/apps` rate limit. When a rate limit violation caused by Terraform Okta provider happens your users start to see errors because some of your app's functionality that relies on `/api/v1/apps` stops working until the rate limit resets.

### Proposal

I suggest we add a new provider config argument https://registry.terraform.io/providers/oktadeveloper/okta/latest/docs#argument-reference:

> `max_requests` - (Optional) (Experimental) controls how many requests can be made to each Okta endpoint by the provider. It's used to prevent rate limit violations. By default request throttling is disabled meaning provider might cause rate limits violations. Expects an integer representing a percentage value - e.g. `40`. `40` means that provider is allowed to use up to 40% of the rate limit. E.g. assuming rate limit for `/api/v1/apps` endpoint is 25, up to 10 requests will be made that burn `/api/v1/apps` rate limit. Currently request throttling works only for `/api/v1/apps` rate limit.

I'd mark this feature as "Experimental" for 2 months to get some feedback from early adopters. I know for sure my organization will start using this feature on day one.

Accepting a signle percentage value is the easiest feature to implement. A more advanced config could look like this:
```
{
	"/api/v1/apps": 40,
	"/api/v1/apps/{id}": 100,
	"/api/v1/groups/{id}": 80
}
```

### Solution

I propose we start with only supporting throttling of requests that burn the `/api/v1/apps` rate limit. It's by far the most pressing problem because `/api/v1/apps` rate limit is significantly lower than any other endpoint.

I think we should use http client's transport to throttle the requests. I propose we change https://github.com/oktadeveloper/terraform-provider-okta/blob/b563d6cbbb2d9e69f76f2ca2ed0ae5a217f221ef/okta/config.go#L48 to:
```
	httpClient.Transport = logging.NewTransport("Okta", httpClient.Transport)
	if c.max_requests < 100 {
		httpClient.Transport = NewRequestThrottleTransport(httpClient.Transport, c.max_requests)
	}
```
RequestThrottleTransport pseudocode:
```
type RequestThrottleTransport struct {
	base http.RoundTripper
	maxRequests int
}
func (throttleTransport *RequestThrottleTransport) RoundTrip(req *http.Request) (res *http.Response, err error) {
	// Check if req uses one of the supported endpoints that we know how throttle.
	// Check how many requests were made during current rate limit interval.
	// Do time.Sleep(...) before making the request if needed.
	resp, err := throttleTransport.base.RoundTrip(req)
	// check response headers, save relevant information
	return resp, err
}
```

Please check https://github.com/oktadeveloper/terraform-provider-okta/pull/187

### Problem with rate limit docs

https://developer.okta.com/docs/reference/rl-global-mgmt/ rate limit documentation is really confusing. By reading the docs you'd never guess that these requests:
1. GET /api/v1/apps/{app_id}/groups?limit=200
2. GET /api/v1/apps/{app_id}/users?limit=200
3. GET /api/v1/apps/{app_id}/groups/{group_id}

share the same rate limit with `/api/v1/apps` endpoint. It will be very hard to implement correct rate limit handling of rate limits other than `/api/v1/apps` for community members that aren't Okta employees.

### FAQ
1. > Does `backoff`, `min_wait_seconds` and `max_wait_seconds` solve your issue?

    No because these settings control the provider's behavior after the rate limit violation happened. We want to avoid rate limit violation from happening at all cost.
2. > Can't you use `-parellelism=1` as in `terraform apply -parellelism=1` to mitigate this issue?

    Even with `-parellelism=1` enough requests are made to Okta to cause a rate limit violation.
3. > What about `sleep()`? Does that help?

    sleep will have no inpact on the `plan` phase which causes the rate limit violations.
4. > Is there any workaround that Terraform Okta provider users can use right now?

    You have to either maintain a custom fork or use a proxy that slows down requests made to Okta. With proxy you have to be really careful to not cause timeouts and if you are a Terraform Cloud user you can't use the proxy workaround.

5. > Latest Golang Okta SDK offers request retries on 429 response status code. Surely that solves your problem, right?

    It doesn't help us at all. We can't afford to have the rate limit completely exhausted by terraform-provider-okta. If terraform-provider-okta uses the whole rate limit allowance our app will become unusable until the rate limit resets. We want to tell terraform-provider-okta to only use up to 25% of the rate limit and leave the remaining 75% for our app.

### PR DESCRIPTION
**Pull request by monde**
@bogdanprodan-okta @adw1n @programmer04 @eparisca @peteryaworsky-okta
This is a rework of PR #187 but using a percentage of the given API capacity for the Okta API management endpoints.

I found how the AWS provider is using a global mutex to protect too many calls going to AWS security groups API and took their approach here. @peteryaworsky-okta suggested we try this capacity percentage governor/throttle approach when we discussed API rate limiting with his team.

I need to bring tests into the PR on the transport as @adw1n original did in #187 but wanted to get this on everyone's radar first.

- https://github.com/hashicorp/terraform-provider-aws
- https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/provider.go#L1576
- https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/internal/mutexkv/mutexkv.go

### CODE REVIEW

**Comment by monde** [comment]
I need to touch up 429 retry logic in addition to bringing in http/transport tests.

**Comment by monde** [comment]
> I need to touch up 429 retry logic in addition to bringing in http/transport tests.

governed transport can run on the retry client or the default client, this is good to go as is.

**Review by bogdanprodan-okta — APPROVED** [review]

File: okta/internal/apimutex/apimutex.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+	// if the status hasn't been updated recently assume there is capacity
+	now := time.Now().Unix()
+	if status.reset+60 < now {
+		return true
+	}
+
+	// calculate utilization
+	utilization := 100.0 * (float32(status.limit-status.remaining) / float32(status.limit))
+	fmt.Println(utilization)
```
Comment (monde) [inline okta/internal/apimutex/apimutex.go]: I must have left that while println debugging, will delete

File: okta/internal/apimutex/apimutex.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		result = "apps"
+	case strings.HasPrefix(endPoint, "/api/v1/users"):
+		result = "users"
+	default:
+		result = "other"
+	}
+	return result
+}
+
+func (s *ApiStatus) Reset() int64 {
```
Comment (monde) [inline okta/internal/apimutex/apimutex.go:117]: need to document public methods

File: okta/internal/transport/governed_transport.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	X_RATE_LIMIT_REMAINING = "x-rate-limit-remaining"
+	X_RATE_LIMIT_RESET     = "x-rate-limit-reset"
+)
+
+type GovernedTransport struct {
+	base     http.RoundTripper
+	apiMutex *apimutex.ApiMutex
+}
+
+// newRequestThrottleTransport returns RoundTripper which provides throttling according to maxRequests.
```
Comment (monde) [inline okta/internal/transport/governed_transport.go]: document public methods and fix typos

**Pull request merged**